core/src/orderbook/streamed/block_timestamp_reading.rs

```diff
@@ -74,59 +82,199 @@ async fn query_block_timestamps_batched(
                 block
                     .hash
                     .expect("blocks queried by hash should contain a hash"),
-                block.timestamp.low_u64(),
+                block,
             ))
         })
         .collect()
 }
 
+/// A block cache ID. This is different than a `BlockId` in that it can't take
+/// special values like `latest` and `earliest` and it implements `Hash` so it
+/// can be use to index into a `HashMap`.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+enum BlockCacheId {
+    Hash(H256),
+    Number(u64),
+}
+
+impl TryFrom<BlockId> for BlockCacheId {
+    type Error = BlockId;
+
+    fn try_from(block_id: BlockId) -> Result<Self, Self::Error> {
+        match block_id {
+            BlockId::Hash(hash) => Ok(BlockCacheId::Hash(hash)),
+            BlockId::Number(BlockNumber::Number(number)) => {
+                Ok(BlockCacheId::Number(number.as_u64()))
+            }
+            _ => Err(block_id),
+        }
+    }
+}
+
+impl Into<BlockId> for BlockCacheId {
+    fn into(self) -> BlockId {
+        match self {
+            BlockCacheId::Hash(hash) => BlockId::Hash(hash),
+            BlockCacheId::Number(number) => BlockId::Number(number.into()),
+        }
+    }
+}
+
 /// A cache for the block timestamp.
 #[derive(Debug)]
 pub struct CachedBlockTimestampReader<T> {
     inner: T,
-    cache: HashMap<H256, u64>,
+    latest_block: u64,
+    confirmation_count: u64,
+    cache: HashMap<BlockCacheId, u64>,
 }
 
-impl<T: BlockTimestampBatchReading + Send> CachedBlockTimestampReader<T> {
-    pub fn new(inner: T) -> Self {
+impl<T> CachedBlockTimestampReader<T> {
+    pub fn new(inner: T, confirmation_count: u64) -> Self {
         Self {
             inner,
+            latest_block: 0,
+            confirmation_count,
             cache: HashMap::new(),
         }
     }
 
+    pub fn update_latest_block(&mut self, latest_block_number: u64) {
```

Comment:
The `prepare_cache` calls it internally, since it needs to be called **before** `prepare_cache` and I didn't want any implicit requirements on the order of calling these methods.

I left it public in case there are uses where there is no `prepare_cache` call but still want to update the `latest_block`.